workorder-job-repeater.lua

```diff
@@ -0,0 +1,278 @@
+        if o.id == id then
+            return o
+        end
+    end
+    log(WARN, "Couldn't find work order", id)
+    return nil
+end
+
+local findBuildingById = function (id)
+    return df.building.find(id)
+end
+
+local findJobById = function (id)
+    for _link, j in utils.listpairs(df.global.world.jobs.list) do
+        if j.id == id then
+            return j
+        end
+    end
+    return nil
+end
+
+local only_infty = true -- we will only repeat jobs which work order's amount is set to infinity. Or not.
+
+-- from http://www.bay12forums.com/smf/index.php?topic=164123.msg8300009#msg8300009
+--[[ "I don't fully understand how this works, so this might be horribly wrong
+      The DF job assigner needs a job to be in postings for it to automatically be assigned to a unit."(c)
+--]]
+local function addJobToPostings(job)
+    local addedIndex = false
+    -- Find the first free empty postings to add a job to
+    for index, posting in ipairs(df.global.world.jobs.postings) do
+        if posting.job == nil then
+            -- It's free real estate!
+            posting.job = job
+            posting.anon_1 = 0
+            posting.flags.dead = false
+
+            job.posting_index = posting.idx
+            addedIndex = posting.idx
+            break
+        end
+    end
+    if addedIndex then
+        log(DEBUG, "Added job " .. job.id .. " to postings as index " .. addedIndex)
+    else
+        log(DEBUG, "Couldn't find posting slot to add in job " .. job.id)
+    end
+
+    return addedIndex
+end
+
+local function repeatJob(order, job)
+    if not (job and order) then
+        -- should never ever happen #hope
+        qerror("job or order is nil")
+        return
+    end
+
+    if only_infty and 0 ~= order.amount_total then
+        log(DEBUG, "skipping job " .. dfhack.job.getName(job) .. " (id " .. job.id .. "): order.amount_total > 0")
+        return
+    end
+
+    if not (order.status.validated and order.status.active) then
+        log(DEBUG, "skipping job " .. dfhack.job.getName(job) .. " (id " .. job.id .. ") from order " .. order.id .. " because of order.status")
+        return
+    end
+
+    -- for debug purposes: pause the game
+    --df.global.pause_state = true
+
+    log(DEBUG, "Repeat job " .. dfhack.job.getName(job) .. " (id " .. job.id .. ") from order " .. order.id)
+
+    -- we need a clone because original(which itself is a copy) is destroyed or something.
+    -- cloneJobStruct doesn't copy flags over.
+    local by_manager = job.flags.by_manager
+    local do_now = job.flags.do_now
+    job = dfhack.job.cloneJobStruct(job)
+    job.flags.by_manager = by_manager
+    job.flags.do_now = do_now
+
+    -- step 0: adjust what we can right now
+    job.flags.working = false
+    job.completion_timer = -1
+
+    -- remove old items we worked upon
+    log(DEBUG, "Freeing old items")
+    for _, item in pairs(job.items) do
+        item:delete()
+    end
+    job.items:resize(0)
```

Comment:
just to check, are there ever any items in this vector? https://github.com/DFHack/dfhack/blob/develop/library/modules/Job.cpp#L89 clears the item and specific refs if `keepEverything` is false (which is the default). Is `keepEverything` somehow getting passed as `true`?

Reply:
I still have this question